File: Windows/Chronicy/Chronicy.Standard/Web/Converters/CardConverter.cs
using Chronicy.Data;
using Chronicy.Utils;
using System;

namespace Chronicy.Web.Converters
{
    public class CardConverter : IConverter<Data.Card, Web.Models.Card>
    {
        private readonly Lazy<CustomFieldConverter> fieldConverter = new Lazy<CustomFieldConverter>();
        private readonly Lazy<TagConverter> tagConverter = new Lazy<TagConverter>();

        public bool CanReverseConvert => true;

        public Models.Card Convert(Data.Card value)
        {
            return new Models.Card
            {
                ID = value.ID,
                Name = value.Name,
                Comment = value.Comment,
                Fields = DataUtils.ValueOrDefault(value.Fields).ConvertAll(input => fieldConverter.Value.Convert(input)),
                Tags = DataUtils.ValueOrDefault(value.Tags).ConvertAll((input) => tagConverter.Value.Convert(input))
            };
        }

        public Data.Card ReverseConvert(Models.Card value)
        {
            return new Data.Card(value.Name)
            {
                ID = value.ID,
                Comment = value.Comment,
                Fields = DataUtils.ValueOrDefault(value.Fields).ConvertAll(input => fieldConverter.Value.ReverseConvert(input)),
                Tags = DataUtils.ValueOrDefault(value.Tags).ConvertAll((input) => tagConverter.Value.ReverseConvert(input))
            };
        }
    }

    public static class CardConverterExtensions
    {
        public static Models.Card ToWebCard(this Data.Card card)
        {
            CardConverter converter = new CardConverter();
            return converter.Convert(card);
        }

        public static Data.Card ToDataCard(this Models.Card card)
        {
            CardConverter converter = new CardConverter();
            return converter.ReverseConvert(card);
        }
    }
}

File: macOS/ChronicyFramework/Shared/Web/Converters/CardConverter.swift
//
//  CardConverter.swift
//  ChronicyFrameworkMacOS
//
//  Created by Alexandru Istrate on 25/07/2019.
//

import Foundation;

public class CardConverter: Converter<Card, CardModel> {
    private let fieldConverter: CustomFieldConverter = CustomFieldConverter();
    private let tagConverter: TagConverter = TagConverter();
    
    public override var allowsReverseConvert: Bool { return true; }
    
    public override func convert(item: Card) -> CardModel {
        let result: CardModel = CardModel();
//        result.id = item.id;
        result.name = item.name;
        result.comment = item.comment;
        result.date = item.date;
        result.fields = item.fields.map({ (iter: CustomField) -> CustomFieldModel in
            return fieldConverter.convert(item: iter);
        });
        result.tags = item.tags.map({ (iter: CardTag) -> TagModel in
            return tagConverter.convert(item: iter);
        });
        return result;
    }
    
    public override func reverseConvert(item: CardModel) -> Card {
        let result: Card = Card(title: item.name, comment: item.comment, date: item.date);
//        result.id = item.id;
        result.fields = item.fields.map({ (iter: CustomFieldModel) -> CustomField in
            return fieldConverter.reverseConvert(item: iter);
        });
        result.tags = item.tags.map({ (iter: TagModel) -> CardTag in
            return tagConverter.reverseConvert(item: iter);
        });
        return result;
    }
}
